OpenRA.Mods.Common/Traits/Player/BulkProductionQueue.cs

```diff
@@ -0,0 +1,361 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Attach this to the player actor (not a building!) to define a new shared build queue.",
+		"Will only work together with the StartProductionProduction: trait on the actor that actually does the production.",
+		"You will also want to add PrimaryBuildings: to let the user choose where new units should exit.")]
+	public class BulkProductionQueueInfo : ProductionQueueInfo, Requires<TechTreeInfo>, Requires<PlayerResourcesInfo>
+	{
+		[Desc("Maximum deliver capacity")]
+		public readonly int MaxCapacity = 6;
+
+		[Desc("delivery delay in ticks")]
+		public readonly int DeliveryDelay = 1500;
+
+		[Desc("Notification played when deliver started")]
+		public readonly string StartDeliveryAudio = null;
+
+		[Desc("Refund Actors that wasn't delivered")]
+		public readonly bool RefundUndeliveredActors = true;
+
+		[Desc("Notification displayed when deliver started")]
+		public readonly string StartDeliveryNotification = null;
+
+		[Desc("Notification played while frigate is on the way. Starting with fist. Last is when delivery actor is spawned on the map.")]
+		public readonly string[] DeliveryProgressNotifications = Array.Empty<string>();
+
+		public override object Create(ActorInitializer init) { return new BulkProductionQueue(init, this); }
+	}
+
+	public class BulkProductionQueue : ProductionQueue
+	{
+		static readonly ActorInfo[] NoItems = Array.Empty<ActorInfo>();
+
+		readonly Actor self;
+		readonly BulkProductionQueueInfo info;
+
+		// Used in refunds. Key is resources, Value is cash
+		readonly List<KeyValuePair<int, int>> actorsCost = new();
+		protected readonly List<ActorInfo> ActorsReadyForDelivery = new();
```

Comment:
```suggestion
		protected readonly List<(ActorInfo Actor, int Resources, int Cash)> ActorsReadyForDelivery = new();
```

just use a tuple. Then it will both result in cleaner and more readable code, and we won't use the ridiculously expensive IndexOf operation.

This also poses issues for the future resource rework, but we will figure that out when we get to it